# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [48]:
import pandas as pd
import numpy as np

da_ma=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')
da_ma.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [49]:
shape=da_ma.shape
print(f'dataset has {shape[0]} rows and {shape[1]} columns')
check_dup=da_ma.duplicated()
print (f'the data set has {check_dup.sum()} duplicated rows')

dataset has 10910 rows and 26 columns
the data set has 0 duplicated rows


In [50]:
da_ma.isna().sum()

Unnamed: 0                          0
Customer                            0
State                             631
Customer Lifetime Value             0
Response                          631
Coverage                            0
Education                           0
Effective To Date                   0
EmploymentStatus                    0
Gender                              0
Income                              0
Location Code                       0
Marital Status                      0
Monthly Premium Auto                0
Months Since Last Claim           633
Months Since Policy Inception       0
Number of Open Complaints         633
Number of Policies                  0
Policy Type                         0
Policy                              0
Renew Offer Type                    0
Sales Channel                       0
Total Claim Amount                  0
Vehicle Class                     622
Vehicle Size                      622
Vehicle Type                     5482
dtype: int64

In [51]:
da_ma.columns=da_ma.columns.str.strip().str.lower().str.replace(' ','_')
da_ma.columns

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type'],
      dtype='object')

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [52]:
# I was not sure If I understood the question correctly therfore:
# if it means to have all the columns but only with the costumer who total_claim_amount > 1000& response = Yes
da_claim= da_ma[(da_ma.total_claim_amount > 1000)& (da_ma.response == 'Yes')]
da_claim.head()

unnamed:_0 customer       state  customer_lifetime_value response  \
189         189  OK31456  California             11009.130490      Yes   
236         236  YJ16163      Oregon             11009.130490      Yes   
419         419  GW43195      Oregon             25807.063000      Yes   
442         442  IP94270     Arizona             13736.132500      Yes   
587         587  FJ28407  California              5619.689084      Yes   

     coverage             education effective_to_date employmentstatus gender  \
189   Premium              Bachelor           1/24/11         Employed      F   
236   Premium              Bachelor           1/24/11         Employed      F   
419  Extended               College           2/13/11         Employed      F   
442   Premium                Master           2/13/11         Disabled      F   
587   Premium  High School or Below           1/26/11       Unemployed      M   

     ...  number_of_open_complaints number_of_policies     policy_type  \
189  ...                        0.0                  1  Corporate Auto   
236  ...                        0.0                  1    Special Auto   
419  ...                        1.0                  2   Personal Auto   
442  ...                        0.0                  8   Personal Auto   
587  ...                        0.0                  1   Personal Auto   

           policy  renew_offer_type  sales_channel  total_claim_amount  \
189  Corporate L3            Offer2          Agent         1358.400000   
236    Special L3            Offer2          Agent         1358.400000   
419   Personal L2            Offer1         Branch         1027.200000   
442   Personal L2            Offer1            Web         1261.319869   
587   Personal L1            Offer2            Web         1027.000029   

     vehicle_class vehicle_size vehicle_type  
189     Luxury Car      Medsize          NaN  
236     Luxury Car      Medsize            A  
419     Luxury Car        Small            A  
442            SUV      Medsize            A  
587            SUV      Medsize            A  

[5 rows x 26 columns]

In [53]:
# if it means to only have total_claim_amount and responce columns for  > 1000 and Yes
# then I drop the other columns : 
column_to_drop=['unnamed:_0', 'customer', 'state', 'customer_lifetime_value', 'coverage', 'education', 'effective_to_date','employmentstatus', 'gender', 'income', 'location_code','marital_status', 'monthly_premium_auto', 'months_since_last_claim','months_since_policy_inception', 'number_of_open_complaints','number_of_policies', 'policy_type', 'policy', 'renew_offer_type','sales_channel', 'vehicle_class', 'vehicle_size','vehicle_type']
da_claim2 =da_claim.drop(columns=column_to_drop)
da_claim2.head(4)

response  total_claim_amount
189      Yes         1358.400000
236      Yes         1358.400000
419      Yes         1027.200000
442      Yes         1261.319869

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [59]:
da_yes= da_ma[(da_ma.response=='Yes')]
da_yes.head(4)
da_yes_filter= da_yes.pivot_table(index='gender', columns='policy_type', values='total_claim_amount',aggfunc='mean').round(2)
display(da_yes_filter)

policy_type  Corporate Auto  Personal Auto  Special Auto
gender                                                  
F                    433.74         452.97        453.28
M                    408.58         457.01        429.53

1- For females, the highest average total claim amount belongs to the "Special Auto" policy type, followed closely by the "Personal Auto" policy type while "Corporate Auto" policy type has the lowest total claim amount.For males, the highest average total claim amount is observed for the "Personal Auto" followed by the "Corporate Auto" policy type while the "Special Auto" policy type has the lowest claim amount.
2- Overall, males tend to have a slightly lower average total claim amount except for the "Personal Auto".

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [61]:
da_st = da_ma.groupby('state')
da_st.size()
# as all the states have more than 400 costumers, so I guess the question here did not mean the original dataset anymore

state
Arizona       1937
California    3552
Nevada         993
Oregon        2909
Washington     888
dtype: int64

In [62]:
# so in dataset with yes response from question 2:
da_st2 = da_yes.groupby('state')
da_st2.size()
# in this case it would be only California

state
Arizona       281
California    504
Nevada        140
Oregon        419
Washington    122
dtype: int64

In [63]:
da_california=da_yes[da_yes.state=='California']
da_california.head()



unnamed:_0 customer       state  customer_lifetime_value response  \
8             8  FM55990  California              5989.773931      Yes   
15           15  CW49887  California              4626.801093      Yes   
19           19  NJ54277  California              3746.751625      Yes   
122         122  EW33419  California             15114.402410      Yes   
125         125  BM92346  California              3919.366722      Yes   

     coverage             education effective_to_date employmentstatus gender  \
8     Premium               College           1/19/11         Employed      M   
15      Basic                Master           1/16/11         Employed      F   
19   Extended               College           2/26/11         Employed      F   
122     Basic  High School or Below           1/19/11    Medical Leave      F   
125  Extended  High School or Below            1/5/11         Employed      F   

     ...  number_of_open_complaints number_of_policies     policy_type  \
8    ...                        0.0                  1   Personal Auto   
15   ...                        0.0                  1    Special Auto   
19   ...                        1.0                  1   Personal Auto   
122  ...                        1.0                  2  Corporate Auto   
125  ...                        0.0                  1   Personal Auto   

           policy  renew_offer_type  sales_channel  total_claim_amount  \
8     Personal L1            Offer2         Branch          739.200000   
15     Special L1            Offer2         Branch          547.200000   
19    Personal L2            Offer2    Call Center           19.575683   
122  Corporate L2            Offer2    Call Center          480.000000   
125   Personal L1            Offer3          Agent          558.099357   

     vehicle_class vehicle_size vehicle_type  
8       Sports Car      Medsize          NaN  
15             SUV      Medsize          NaN  
19    Two-Door Car        Large            A  
122            SUV      Medsize            A  
125  Four-Door Car        Small          NaN  

[5 rows x 26 columns]

In [64]:
# I could also use get_group method. 
da_st2.get_group('California').head()

unnamed:_0 customer       state  customer_lifetime_value response  \
8             8  FM55990  California              5989.773931      Yes   
15           15  CW49887  California              4626.801093      Yes   
19           19  NJ54277  California              3746.751625      Yes   
122         122  EW33419  California             15114.402410      Yes   
125         125  BM92346  California              3919.366722      Yes   

     coverage             education effective_to_date employmentstatus gender  \
8     Premium               College           1/19/11         Employed      M   
15      Basic                Master           1/16/11         Employed      F   
19   Extended               College           2/26/11         Employed      F   
122     Basic  High School or Below           1/19/11    Medical Leave      F   
125  Extended  High School or Below            1/5/11         Employed      F   

     ...  number_of_open_complaints number_of_policies     policy_type  \
8    ...                        0.0                  1   Personal Auto   
15   ...                        0.0                  1    Special Auto   
19   ...                        1.0                  1   Personal Auto   
122  ...                        1.0                  2  Corporate Auto   
125  ...                        0.0                  1   Personal Auto   

           policy  renew_offer_type  sales_channel  total_claim_amount  \
8     Personal L1            Offer2         Branch          739.200000   
15     Special L1            Offer2         Branch          547.200000   
19    Personal L2            Offer2    Call Center           19.575683   
122  Corporate L2            Offer2    Call Center          480.000000   
125   Personal L1            Offer3          Agent          558.099357   

     vehicle_class vehicle_size vehicle_type  
8       Sports Car      Medsize          NaN  
15             SUV      Medsize          NaN  
19    Two-Door Car        Large            A  
122            SUV      Medsize            A  
125  Four-Door Car        Small          NaN  

[5 rows x 26 columns]

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [66]:
da_california.groupby(['gender', 'education'])['customer_lifetime_value'].agg(['max', 'min', 'median'])

max          min        median
gender education                                                    
F      Bachelor              33473.349460  2248.449633   5217.513287
       College               25807.063000  2004.350666   6273.911874
       Doctor                 6265.343299  2395.570000   4014.453113
       High School or Below  41787.903430  2397.036098   5979.724161
       Master                19160.989940  2574.020376   4861.737158
M      Bachelor              24127.504020  2300.691547   4984.618655
       College               22332.439460  2393.915369   6005.847375
       Doctor                12731.951610  3436.134281  10972.067780
       High School or Below  12298.686300  2227.072755   6366.225775
       Master                14435.673650  2551.226692   8509.850887

Females with a High School or Below degree have the maximum customer lifetime value, indicating the potential for this demographic. However, the minimum value is 2248.45, suggesting a wide range of customer value within this group. The median customer eflects the central tendency of values.In male clients however  the maximum customer lifetime value belongs to those who have a Bechelor's degree and those with a highschool or below have the lowest maximum in this Demographic 12298.6.

There is significant variability within each education level, as evidenced by the range between the maximum and minimum values. For example, females with a Master's degree have a maximum value of 19,160.99, whereas the minimum value is 2,574.02. Therefore the factors beyond education level play an important role.

Across all education levels except "Doctorate", females tend to have higher maximum customer lifetime values compared to males. For instance, females with a Bachelor's degree have a maximum value of 33,473.35, while males with the same education level have a maximum value of 24,127.50. 

The median customer lifetime value provides insights into the central tendency of values within each group. While females generally have higher maximum values, males have slightly higher median values in some education levels but it is particularly evident among those with a Doctorate degree.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here